src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/PartitionedRateLimiter.cs

```diff
@@ -182,48 +184,77 @@ private void ThrowIfDisposed()
+                        idleDuration = rateLimiter.Value.Value.IdleDuration ?? TimeSpan.Zero;
+                        if (idleDuration > _idleTimeLimit)
                         {
-                            // In rare cases the RateLimiter will be added to the storage but not be initialized yet
-                            // keep cache invalid if there was a non-initialized RateLimiter
-                            // the next time we run the timer the cache will be updated
-                            // with the initialized RateLimiter
-                            cacheStillInvalid = true;
+                            // Remove limiter from the lookup table and mark cache as invalid
+                            // If a request for this partition comes in it will have to create a new limiter now
+                            // And the next time the timer runs the cache needs to be updated to no longer have a reference to this limiter
+                            _cacheInvalid = true;
+                            _limiters.Remove(rateLimiter.Key);
+
+                            // We don't want to dispose inside the lock so we need to defer it
+                            limitersToDispose ??= new List<RateLimiter>();
+                            limitersToDispose.Add(rateLimiter.Value.Value);
                         }
                     }
-                    limiter._cacheInvalid = cacheStillInvalid;
+                }
+                else if (rateLimiter.Value.Value is ReplenishingRateLimiter replenishingRateLimiter)
+                {
+                    replenishingRateLimiter.TryReplenish();
                 }
             }
 
-            // cachedLimiters is safe to use outside the lock because it is only updated by the Timer
-            // and the Timer avoids re-entrancy issues via the _executingTimer field
-            foreach (Lazy<RateLimiter> rateLimiter in limiter._cachedLimiters)
+            if (limitersToDispose is not null)
             {
-                Debug.Assert(rateLimiter.IsValueCreated && rateLimiter.Value is ReplenishingRateLimiter);
-                ((ReplenishingRateLimiter)rateLimiter.Value).TryReplenish();
+                List<Exception>? aggregateExceptions = null;
+                foreach (RateLimiter limiter in limitersToDispose)
+                {
+                    try
+                    {
+                        await limiter.DisposeAsync().ConfigureAwait(false);
+                    }
+                    catch (Exception ex)
+                    {
+                        aggregateExceptions ??= new List<Exception>();
+                        aggregateExceptions.Add(ex);
+                    }
+                }
+                if (aggregateExceptions is not null)
+                {
+                    throw new AggregateException(aggregateExceptions);
```

Comment:
Should we have a try/catch per call to TryReplenish too? We really need to figure out how we're going to handle these errors.

Do we need add an error callback API? Does `PartitionedRateLimiter<TResource>` need its own `TryReplenish` method that doesn't catch?